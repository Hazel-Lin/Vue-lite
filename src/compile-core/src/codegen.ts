import { NodeTypes } from "./ast";
import { helperMapName, TO_DISPLAY_STRING } from "./runtimeHelpers";
export const generate = (ast: any) => {
  // { children: [ { type: 3, content: 'hi' } ] } 
  // console.log(ast,'ast')
  // 生成string
  // let code = ''
  let result = genContext()
  const { push } = result

  let functionName = 'render'
  let p = ['_ctx','_cache']
  const signature = p.join(", ");

  let returnStr = "return "
  push(`return function ${functionName}(${signature}) {`)
  push(returnStr +  genNode(ast))
  push('}')
  genFunctionPreamble(ast, result);
  // code += `return function ${functionName}(`
  // code +=  p.join(', ') + ') {'
  // code += returnStr +  genNode(ast)
  // code += '}'
  return {
    code: result.code
  }
}
function genFunctionPreamble(ast, context) {
  const { push } = context;
  const VueBinging = "Vue";
  const aliasHelper = (s) => `${helperMapName[s]}:_${helperMapName[s]}`;
  if (ast.helpers.length > 0) {
    push(
      `const { ${ast.helpers.map(aliasHelper).join(", ")} } = ${VueBinging}`
    );
  }
  push("\n");
  push("return ");
}
function genNode(ast){
  let res = ast.codegenNode
  return `'${res.content}'`
}
function genContext (){
  const context = {
    code: '',
    push(source){
      context.code += source
    }
  }
  return context
}