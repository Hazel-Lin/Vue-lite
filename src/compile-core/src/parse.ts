import { NodeTypes } from "./ast";

const start = "{{"
const end = "}}"
const enum TagType {
  Start,
  End,
}
// ast 是一个对象 包含type和children属性
// children是一个数组 包含子节点
// children[0]是一个对象 包含type和content属性
export const baseParse = (content) =>{
  const context = createParserContext(content)
  // 将这些子节点作为根节点的children
  return createRoot(parseChildren(context,[]));
}
// 将内容放置在source中
function createParserContext(content) {
  return {
    source: content,
  };
}
// 解析children 如果一"{{"开始 则为插值 解析插值
// 重点在于如何解析插值
// <div><p>hi</p>{{message}}</div> 解析元素  解析元素 解析文本 解析插值
// <div>hi,{{message}}</div>
// 1. 循环判断属于哪个类型
function parseChildren(context,ancestors) {
  const nodes:any[] = []
  // console.log('isEnd(context)',isEnd(context));
  while(!isEnd(context,ancestors)){
    let node;
    // context = { source: '{{ message }}' }
    if(context.source.startsWith(start)){
      node = parseInterpolation(context)
    }else if(context.source.startsWith('<')){
      // 如果第一位是< 则解析element 再判断是否是字母
      // console.log(context,'context1');
      // { source: '<div></div>' } context
      if (/[a-z]/i.test(context.source[1])) {
        node = parseElement(context,ancestors);
      }
      // { type: 2, tag: 'div' }
    }
    // console.log(node,'node');

    // 解析文本
    if (!node) {
      node = parseText(context);
    }
    
    nodes.push(node)
    // [ { type: 0, content: { type: 1, content: 'message' } } ]
    // console.log(nodes,'nodes');
  }
  return nodes
}
  // 解析文本
function parseText(context){
  // console.log('context',context.source,context.source.length);
  // 获取到数据 删除空格
  // 删除获取后的代码
  // 直接返回文本节点内容
  // const content = context.source
  let endTokens = ['<','{{']
  // let endTokens = '{{'


  let endIndex = context.source.length
  // 当同时遇到{{或<时 比较两者的index值 取最小的
  // console.log(context.source,'context.source');

  for (let i = 0; i < endTokens.length; i++) {
    const index = context.source.indexOf(endTokens[i]);

    if (index !== -1 && endIndex > index) {
      endIndex = index
    }
  }
  // console.log(endIndex,'endIndex');

  // if(context.source.indexOf(endTokens) > -1){
  //   endIndex = context.source.indexOf(endTokens)
  // }
  const content = parseTextData(context, endIndex);
  // console.log('content---------',content);

  return {
    type: NodeTypes.TEXT,
    content,
  };
}
function parseElement(context,ancestors) {
  // 处理开始标签 返回tag节点
  const element:any = parseTag(context, TagType.Start)
  ancestors.push(element)

  // 递归调用parseChildren
  element.children = parseChildren(context,ancestors)
  // console.log(element,'element');
  ancestors.pop();

  if (startsWithEndTagOpen(context.source, element.tag)) {
    // 处理结束标签
    parseTag(context, TagType.End);
  } else {
    throw new Error(`缺少结束标签:${element.tag}`);
  }
  return element;
}
// 设置一个全局变量
// let parseEndTag:any = null;
// 此时context应该是 <div></div>
function parseTag(context,type:TagType){
  // console.log('context2',context);
  // <div>hi,{{message}}</div>

  const match:any = /^<\/?([a-z]*)/i.exec(context.source)
  // console.log('match-----',match);
  // parseEndTag = match[1]
  const tag = match[1]
  // 删除处理完成的标签
  advanceBy(context, match[0].length);
  advanceBy(context, 1);
  // console.log('context.source',context.source);
  // 处理结束标签时 不需要返回节点
  if (type === TagType.End) return;
  return {
    type: NodeTypes.ELEMENT,
    tag,
  };
}
// 解析插值 '{{ message }}'
function parseInterpolation(context) {
  // 2
  const sl = start.length
  const closeIndex = context.source.indexOf(
    end,
    sl
  );
  if (closeIndex < 0) {
    console.error('插值缺少结束定界符')
  }
  // 11 获取到结束括号 }} 开始的位置
  // console.log('closeIndex',closeIndex);
  // console.log('context',context);
  // 向前推进2位 ' message }}'
  advanceBy(context, sl);
  // 获取到插值的内容 message 的长度 可能存在空格
  const rawContentLength = closeIndex - sl;
  // console.log('context.source',context.source);
    //  context.source =  message }}
  const rawContent = context.source.slice(0, rawContentLength);
  // const rawContent = context.source.slice(sl,closeIndex)
  const content = rawContent.trim()
  // 插值取值完成后 删除插值的内容
  advanceBy(context, rawContentLength + end.length);

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content,
    },
  };
}
function advanceBy(context: any, length: number) {
  context.source = context.source.slice(length);
}
function createRoot(children) {
  return {
    children,
    type: NodeTypes.ROOT
  };
}
// <div><span></div> 只要在数组中找到和结束标签相同的标签就结束 并且报错
// 每一次parseElement的时候都将tag标签存储在一个数组中
function isEnd(context: any, ancestors) {
  // 1. 当source为空时 返回true
  // 2. 当遇到结束标签时 返回true
  const s = context.source;
  if(s.startsWith('</')){
    for(let i = ancestors.length -1 ; i >= 0; i--){
      let tag  = ancestors[i].tag;
      if (startsWithEndTagOpen(s, tag)) {
        return true;
     }
    }
  }
  // if(s.startsWith(`</${ancestors}>`)){
  //   return true;
  // }
 return !s
}
function startsWithEndTagOpen(source, tag) {
  return (
    source.startsWith("</") &&
    source.slice(2, 2 + tag.length).toLowerCase() === tag.toLowerCase()
  );
}
function parseTextData(context: any, length: any) {
  const content = context.source.slice(0, length);
  // 2. 推进
  advanceBy(context, length);
  return content;
}