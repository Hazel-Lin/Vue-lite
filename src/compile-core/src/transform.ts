import { NodeTypes } from "./ast";
import { TO_DISPLAY_STRING } from "./runtimeHelpers";
// 1. 找到对应节点  深度优先
// 2. 修改节点内容
export const transform = (root, options={}) => {
  console.log('transform',root);
  const context:any = createTransformContext(root, options);
  console.log('context',context);
  traverseNode(root, context);
  createRootChildren(root);
  root.helpers = [...context.helpers.keys()];
}
function createRootChildren(root: any) {
 return root.codegenNode = root.children[0];
}

function createTransformContext(root: any, options: any) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper(key) {
      context.helpers.set(key, 1);
    },
  };

  return context;
}
function traverseNode(node: any, context: any) {
  console.log('node',node);

  const nodeTransforms = context.nodeTransforms;
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i];
    // 递归调用transform函数
    transform(node);
  }

  traverseChildren(node, context);
}
function traverseChildren(node: any, context: any) {
  const children = node.children;

  if (children) {
    for (let i = 0; i < children.length; i++) {
      const node = children[i];
      traverseNode(node, context);
    }
  }
}

