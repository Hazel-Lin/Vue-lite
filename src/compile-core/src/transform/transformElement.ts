import { NodeTypes } from "../ast";
import { CREATE_ELEMENT_VNODE } from "../runtimeHelpers";

export const transformElement = function(node,context){
    if(node.type === NodeTypes.ELEMENT){
      return () => {
        context.helper(CREATE_ELEMENT_VNODE)
        const { props,tag,children } = node;
        const vnodeTag = `"${tag}"`;
        const vnodeProps = props;
        let vnodeChildren = children[0];
        const vnodeElement = {
          type: NodeTypes.ELEMENT,
          tag:vnodeTag,
          props:vnodeProps,
          children:vnodeChildren
        }
        node.codegenNode = vnodeElement
      }
     
  }
}