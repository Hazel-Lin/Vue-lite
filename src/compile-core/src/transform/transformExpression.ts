import { NodeTypes } from "../ast";

// 处理节点
export function transformExpression(node) {
  if (node.type === NodeTypes.INTERPOLATION) {
    node.content = processExpression(node.content);
  }
}

function processExpression(node: any) {
  node.content = `_ctx.${node.content}`;
  return node;
}