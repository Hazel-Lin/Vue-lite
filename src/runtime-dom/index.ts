import { createRenderer } from '../runtime-core'

// 方法抽离
function createElement(type) {
  return document.createElement(type)
}

function patchProp(el, key, prevVal, newVal) {
  const isOn = (key: string) => /^on[A-Z]/.test(key)
  if (isOn(key)) {
    const event = key.slice(2).toLowerCase()
    el.addEventListener(event, newVal)
  }
  else {
    if (newVal === undefined || newVal === null)
      el.removeAttribute(key)

    else
      el.setAttribute(key, newVal)
  }
}
// 关于不传值没有出现报错的问题 有待理解
function insert(el, parent, anchor) {
  parent.insertBefore(el, anchor || null)
}
function remove(child) {
  const parent = child.parentNode
  if (parent)
    parent.removeChild(child)
}
function setElementText(el, text) {
  el.textContent = text
}
// 创建渲染器
const renderer: any = createRenderer({
  createElement,
  patchProp,
  insert,
  remove,
  setElementText,
})
// 外部调用的createApp方法
export function createApp(...args) {
  return renderer.createApp(...args)
}

export * from '../runtime-core'
