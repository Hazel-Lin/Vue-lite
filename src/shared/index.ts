export const isObject = (val: unknown): val is Record<any, any> =>
  val !== null && typeof val === 'object'

export const isString = (value) => typeof value === "string";
export const extend = Object.assign
// 判断是否需要改变
export const hasChanged = (newValue,oldValue)=> { 
  return !Object.is(newValue,oldValue)
}
export const hasOwn = (val, key) =>
  Object.prototype.hasOwnProperty.call(val, key);