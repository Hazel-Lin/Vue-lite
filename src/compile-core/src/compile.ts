import { generate } from "./codegen";
import { baseParse } from "./parse";
import { transform } from "./transform";
import { transformElement } from "./transform/transformElement";
import { transformExpression } from "./transform/transformExpression";
import { transformText } from "./transform/transformText";

// 将template 编译成 render函数
export function baseCompile(template) {
  const ast: any = baseParse(template);
  transform(ast, {
    nodeTransforms: [transformExpression, transformElement, transformText],
  });

  return generate(ast);
}
