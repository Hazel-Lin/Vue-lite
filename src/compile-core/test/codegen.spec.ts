import { generate } from "../src/codegen";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";
import { transformExpression } from "../src/transform/transformExpression";
describe("codegen", () => {
  // it("string", () => {
  //   const ast = baseParse("hi");
  //   transform(ast);
  //   const { code } = generate(ast);
  //   // 快照测试 string
  //   expect(code).toMatchSnapshot();
  // });
  it.only("interpolation", () => {
    const ast = baseParse("{{message}}");
    transform(ast);
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });
});
