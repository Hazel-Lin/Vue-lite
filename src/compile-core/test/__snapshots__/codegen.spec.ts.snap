// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen interpolation 1`] = `
"return function render(_ctx, _cache) {return '[object Object]'}
return "
`;

exports[`codegen string 1`] = `"return function render(_ctx, _cache) {return 'hi'}"`;
